### PR TITLE
[sram_ctrl/rtl] Timing fixes

### PR DESCRIPTION
**Pull request by GregAC**
I plan to put together a hacky block synthesis for sram_ctrl so I can check I've eliminated all of the feedthroughs that were causing issues. Will work on that tomorrow.

Contributes to resolving #22462.

### CODE REVIEW

**Comment by GregAC** [comment]
Got the block level synth running I've identified another feedthrough via the `error_internal` signal in `tlul_adapter_sram`: https://github.com/lowRISC/opentitan/blob/2feef77d3a04bb4281d70fdeff9c1cdf458b6320/hw/ip/tlul/rtl/tlul_adapter_sram.sv#L311

I think it'll be shallower feedthrough than the ones I've eliminated but best we get rid of them all. I can get rid of it but it will be a little more work, in particular may involve some extra DV work.

**Comment by vogelpi** [comment]
As this second path is expected to be shallower and involves more DV work, should we first try to finalize the fix for the first and see how this impacts the actual chip timing?

**Comment by GregAC** [comment]
In parallel I'm fixing up the CI issues for these current fixes so they should be good to go once they've been reviewed. Happy to trial how that impacts the chip timing alone but I would like to continue with this other fix. Long feedthroughs starting at Ibex going through the crossbar and then through SRAM and all the way back in Ibex generally aren't good. With them gone the implementation flow may find it has more placement freedom and wants to put in fewer buffers giving better QoR. In particular could give us some modest area savings that we'll appreciate having.

**Comment by GregAC** [comment]
After a bit more investigation this new path may not be that much shallower than the previous ones anyway, so I think worth fixing it before doing a new implementation trial.

File: hw/ip/sram_ctrl/rtl/sram_ctrl.sv
Comment (andreaskurth) [inline hw/ip/sram_ctrl/rtl/sram_ctrl.sv:508]: Do we really need to take `tlul_we` or `sram_rmw_in_progress` into account for `key_vaiid` in the escalated case? As far as I can see, in the escalated case `key_valid` is only used to make `prim_ram_1p_scr` grant requests -- can't we make it grant all requests, including reads?

The reason I'm asking is because the change would be much simpler, and it would also mean less code to maintain

File: hw/ip/sram_ctrl/rtl/sram_ctrl.sv
Comment (andreaskurth) [inline hw/ip/sram_ctrl/rtl/sram_ctrl.sv:490]: This change LGTM

File: hw/ip/sram_ctrl/rtl/sram_ctrl.sv
Comment (GregAC) [reply hw/ip/sram_ctrl/rtl/sram_ctrl.sv:508]: I guess it's a defence in depth? With this setup we've both got a TLUL gate and the clearing of `key_valid` from stopping a transaction acting on the SRAM after escalation. If `key_valid` is always true other than when `key_req_pending_q` is set then after an escalation any transaction that reaches the `tlul_sram_adapter` will get acted upon.

I note there's an explicit clone of the escalation signal, with one clone setting the escalated bit (which causes `key_valid` is go low) and the other clone closing the TLUL gate:

https://github.com/lowRISC/opentitan/blob/9f4903e77a8567a289e22d4c788822afb3a89cc6/hw/ip/sram_ctrl/rtl/sram_ctrl.sv#L164-L172

There's some discussion on these PRs: https://github.com/lowRISC/opentitan/pull/11863 and https://github.com/lowRISC/opentitan/pull/17072

There's also the argument we want to change behaviour as little as possible.

Comment (andreaskurth) [reply hw/ip/sram_ctrl/rtl/sram_ctrl.sv:508]: Thanks for the explanation. I think defense in depth and changing behavior as little as possible are sufficient arguments for making the change this way

**Review by andreaskurth — APPROVED** [review]
LGTM based on the discussions in this PR. Thanks @GregAC!

**Comment by vogelpi** [comment]
Please don't merge this yet, AFAIK Greg found another path he is currently working on improving.

**Comment by GregAC** [comment]
Yes I've got two more fixes inbound, currently investigating some regression failures and then will update this PR with them.

**Comment by andreaskurth** [comment]
Since we are agreed on this, we could just merge this one for now and create a follow-up PR for the other improvements (tracked against #22462)?

**Comment by GregAC** [comment]
Regression issues I'm looking at may be related to these changes so I'd prefer to at least understand those before we merge this.

**Comment by GregAC** [comment]
Failures look to be related to my new work so merging this now

**Pull request merged**
